control/mttrie.go

```diff
@@ -145,81 +141,161 @@ func (mtt *mttNode) Add(mt *metricType) {
+// and does gather all the node's descendants if the namespace ends with an asterisk
+func (mtt *mttNode) GetMetrics(ns []string, ver int) ([]*metricType, error) {
+	nodes := []*mttNode{}
+	mts := []*metricType{}
+
+	// search returns all of the nodes fulfilling the 'ns'
+	// even for some of them there is no metric (empty node.mts)
+	nodes = mtt.search(nodes, ns)
+
+	for _, node := range nodes {
+		// choose the queried version of metric types (or the latest if ver < 1)
+		// and concatenate them into a single slice
+		mt, err := getVersion(node.mts, ver)
+		if err != nil {
+			continue
+		}
 		mts = append(mts, mt)
 	}
+	if len(mts) == 0 {
+		return nil, errorMetricNotFound("/"+strings.Join(ns, "/"), ver)
+	}
+	return mts, nil
+}
+
+// GetVersions returns all versions of MTs below the given namespace
+func (mtt *mttNode) GetVersions(ns []string) ([]*metricType, error) {
+	var nodes []*mttNode
+	var mts []*metricType
+
+	nodes = mtt.search(nodes, ns)
+
+	for _, node := range nodes {
+		// concatenates metric types in ALL versions into a single slice
+		for _, mt := range node.mts {
+			mts = append(mts, mt)
+		}
+	}
+	if len(mts) == 0 {
+		return nil, errorMetricNotFound("/" + strings.Join(ns, "/"))
+	}
 	return mts, nil
 }
 
-// walk returns the last leaf / branch present
-// in the trie and the index in the namespace that the last node exists.
+// fetch collects all descendants nodes below the given namespace
+func (mtt *mttNode) fetch(ns []string) []*mttNode {
+	node, err := mtt.find(ns)
+	if err != nil {
+		return nil
+	}
+
+	var children []*mttNode
+
+	if node.mts != nil {
+		children = append(children, node)
+	}
+	if node.children != nil {
+		children = gatherDescendants(children, node)
+	}
+	return children
+}
+
+// walk returns the last leaf / branch present in the trie and the index in the namespace that the last node exists.
+// It is useful e.g. to locate the right place to add new metric type into tree with the given namespace
 func (mtt *mttNode) walk(ns []string) (*mttNode, int) {
 	parent := mtt
 	var pp *mttNode
+
 	for i, n := range ns {
 		if parent.children == nil {
 			return parent, i
 		}
+
 		pp = parent
 		parent = parent.children[n]
+
 		if parent == nil {
 			return pp, i
 		}
 	}
 	return parent, len(ns)
 }
 
+// search returns leaf nodes in the trie below the given namespace
+func (mtt *mttNode) search(nodes []*mttNode, ns []string) []*mttNode {
+	parent := mtt
+	var children []*mttNode
+
+	if parent.children == nil {
+		return nodes
+	}
+	if len(ns) == 1 {
+		// the last element of ns is under searching process
+		switch ns[0] {
+		case "*":
+			// fetch all descendants when wildcard ends namespace
+			children = parent.fetch([]string{})
+		default:
+			children = parent.gatherChildren(ns[0])
+		}
+		nodes = append(nodes, children...)
+		return nodes
+	}
+	children = parent.gatherChildren(ns[0])
+
+	for _, child := range children {
+		nodes = child.search(nodes, ns[1:])
```

Comment:
Does this not just re-assign over the return value, which would end up returning the results from the last child looked at? I feel like maybe I am missing something.